avalon/tools/widgets.py

```diff
@@ -0,0 +1,327 @@
+
+        silo = self.tabText(index)
+        self.silo_changed.emit(silo)
+
+        # Store for the next calls
+        self._previous_tab_index = index
+
+    def clear(self):
+        """Removes all tabs.
+
+        Implemented similar to `QTabWidget.clear()`
+
+        """
+        for i in range(self.count()):
+            self.removeTab(0)
+
+    def set_silos(self, silos):
+
+        current_silo = self.get_current_silo()
+
+        if not silos:
+            # Add an emtpy stub tab to start on.
+            silos = [""]
+
+        # Populate the silos without emitting signals
+        self.blockSignals(True)
+        self.clear()
+        for silo in sorted(silos):
+            self.addTab(silo)
+
+        if self.silo_creatable:
+            # Add the "+" tab
+            self.addTab("+")
+
+        self.set_current_silo(current_silo)
+        self.blockSignals(False)
+
+        # Assume the current index is "fine"
+        self._previous_tab_index = self.currentIndex()
+
+        # Only emit a silo changed signal if the new signal
+        # after refresh is not the same as prior to it (e.g.
+        # when the silo was removed, or alike.)
+        if current_silo != self.get_current_silo():
+            self.currentChanged.emit(self.currentIndex())
+
+    def set_current_silo(self, silo):
+        """Set the active silo by name or index.
+
+        Args:
+            silo (str or int): The silo name or index.
+
+        """
+
+        # Already set
+        if silo == self.get_current_silo():
+            return
+
+        # Otherwise change the silo box to the name
+        for i in range(self.count()):
+            text = self.tabText(i)
+            if text == silo:
+                self.setCurrentIndex(i)
+                break
+
+    def get_current_silo(self):
+        index = self.currentIndex()
+        return self.tabText(index)
+
+    def on_add_silo(self):
+        silo, state = QtWidgets.QInputDialog.getText(self,
+                                                     "Silo name",
+                                                     "Create new silo:")
+        if not state or not silo:
+            return
+
+        self.add_silo(silo)
+
+    def get_silos(self):
+        """Return the currently available silos"""
+
+        # Ignore first tab if empty
+        # Ignore the last tab because it is the "+" tab
+        silos = []
+        for i in range(self.count() - 1):
+            label = self.tabText(i)
+            if i == 0 and not label:
+                continue
+            silos.append(label)
+        return silos
+
+    def add_silo(self, silo):
+
+        # Add the silo
+        silos = self.get_silos()
+        silos.append(silo)
+        silos = list(set(silos))  # ensure unique
+        self.set_silos(silos)
+        self.silo_added.emit(silo)
+
+        self.set_current_silo(silo)
+
+
+def _list_project_silos():
+    """List the silos from the project's configuration"""
+    silos = io.distinct("silo")
```

Comment:
This function and entire `avalon/tools/projectmanager/lib.py` looks like they could be put together into `avalon/tools/database.py` ? A place for database io functions. What do you think ?

Reply:
I'm not necessarily for *more separation* where it's currently not required. Unless it's solving a particular problem. This function is solely there because one of the widgets uses it, as described in the contribution manual it should be as close as possible to the classes/functions using it.

If we would move it into a `database.py` or something alike we could make it a separate PR.

Reply:
Yes, let's keep it contained until it absolutely must be separated. Less indirection the better.